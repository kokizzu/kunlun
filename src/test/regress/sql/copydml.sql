--
-- Test cases for COPY (INSERT/UPDATE/DELETE) TO
--
create table copydml_test (id serial, t text);
insert into copydml_test (t) values ('a');
insert into copydml_test (t) values ('b');
insert into copydml_test (t) values ('c');
insert into copydml_test (t) values ('d');
insert into copydml_test (t) values ('e');

--
-- Test COPY (insert/update/delete ...)
--
copy (insert into copydml_test (t) values ('f') returning id) to stdout;
copy (update copydml_test set t = 'g' where t = 'f' returning id) to stdout;
copy (delete from copydml_test where t = 'g' returning id) to stdout;

--
-- Test \copy (insert/update/delete ...)
--
\copy (insert into copydml_test (t) values ('f') returning id) to stdout;
\copy (update copydml_test set t = 'g' where t = 'f' returning id) to stdout;
\copy (delete from copydml_test where t = 'g' returning id) to stdout;

-- Error cases
copy (insert into copydml_test default values) to stdout;
copy (update copydml_test set t = 'g') to stdout;
copy (delete from copydml_test) to stdout;

copy (insert into copydml_test default values) to stdout;
copy (insert into copydml_test default values) to stdout;
copy (insert into copydml_test default values) to stdout;
copy (insert into copydml_test default values) to stdout;

copy (update copydml_test set t = 'f') to stdout;
copy (update copydml_test set t = 'f') to stdout;
copy (update copydml_test set t = 'f') to stdout;
copy (update copydml_test set t = 'f') to stdout;

copy (delete from copydml_test) to stdout;
copy (delete from copydml_test) to stdout;
copy (delete from copydml_test) to stdout;
copy (delete from copydml_test) to stdout;

-- triggers
create function qqq_trig() returns trigger as $$
begin
if tg_op in ('INSERT', 'UPDATE') then
    raise notice '% %', tg_op, new.id;
    return new;
else
    raise notice '% %', tg_op, old.id;
    return old;
end if;
end
$$ language plpgsql;
create trigger qqqbef before insert or update or delete on copydml_test
    for each row execute procedure qqq_trig();
create trigger qqqaf after insert or update or delete on copydml_test
    for each row execute procedure qqq_trig();
copy (insert into copydml_test (t) values ('f') returning id) to stdout;
copy (update copydml_test set t = 'g' where t = 'f' returning id) to stdout;
copy (delete from copydml_test where t = 'g' returning id) to stdout;

drop table copydml_test;
drop function qqq_trig();						 
