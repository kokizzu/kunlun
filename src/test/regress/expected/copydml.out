--
-- Test cases for COPY (INSERT/UPDATE/DELETE) TO
--
create table copydml_test (id serial primary key, t text);
CREATE TABLE
insert into copydml_test (t) values ('a');
INSERT 0 1
insert into copydml_test (t) values ('b');
INSERT 0 1
insert into copydml_test (t) values ('c');
INSERT 0 1
insert into copydml_test (t) values ('d');
INSERT 0 1
insert into copydml_test (t) values ('e');
INSERT 0 1
--
-- Test COPY (insert/update/delete ...)
--
copy (insert into copydml_test (t) values ('f') returning id) to stdout;
6
copy (update copydml_test set t = 'g' where t = 'f' returning id) to stdout;
6
copy (delete from copydml_test where t = 'g' returning id) to stdout;
6
--
-- Test \copy (insert/update/delete ...)
--
\copy (insert into copydml_test (t) values ('f') returning id) to stdout;
7
\copy (update copydml_test set t = 'g' where t = 'f' returning id) to stdout;
7
\copy (delete from copydml_test where t = 'g' returning id) to stdout;
7
-- Error cases
copy (insert into copydml_test default values) to stdout;
psql:sql/copydml.sql:26: ERROR:  COPY query must have a RETURNING clause
copy (update copydml_test set t = 'g') to stdout;
psql:sql/copydml.sql:27: ERROR:  COPY query must have a RETURNING clause
copy (delete from copydml_test) to stdout;
psql:sql/copydml.sql:28: ERROR:  COPY query must have a RETURNING clause
copy (insert into copydml_test default values) to stdout;
psql:sql/copydml.sql:30: ERROR:  COPY query must have a RETURNING clause
copy (insert into copydml_test default values) to stdout;
psql:sql/copydml.sql:31: ERROR:  COPY query must have a RETURNING clause
copy (insert into copydml_test default values) to stdout;
psql:sql/copydml.sql:32: ERROR:  COPY query must have a RETURNING clause
copy (insert into copydml_test default values) to stdout;
psql:sql/copydml.sql:33: ERROR:  COPY query must have a RETURNING clause
copy (update copydml_test set t = 'f') to stdout;
psql:sql/copydml.sql:35: ERROR:  COPY query must have a RETURNING clause
copy (update copydml_test set t = 'f') to stdout;
psql:sql/copydml.sql:36: ERROR:  COPY query must have a RETURNING clause
copy (update copydml_test set t = 'f') to stdout;
psql:sql/copydml.sql:37: ERROR:  COPY query must have a RETURNING clause
copy (update copydml_test set t = 'f') to stdout;
psql:sql/copydml.sql:38: ERROR:  COPY query must have a RETURNING clause
copy (delete from copydml_test) to stdout;
psql:sql/copydml.sql:40: ERROR:  COPY query must have a RETURNING clause
copy (delete from copydml_test) to stdout;
psql:sql/copydml.sql:41: ERROR:  COPY query must have a RETURNING clause
copy (delete from copydml_test) to stdout;
psql:sql/copydml.sql:42: ERROR:  COPY query must have a RETURNING clause
copy (delete from copydml_test) to stdout;
psql:sql/copydml.sql:43: ERROR:  COPY query must have a RETURNING clause
-- triggers
create function qqq_trig() returns trigger as $$
begin
if tg_op in ('INSERT', 'UPDATE') then
    raise notice '% %', tg_op, new.id;
    return new;
else
    raise notice '% %', tg_op, old.id;
    return old;
end if;
end
$$ language plpgsql;
CREATE FUNCTION
create trigger qqqbef before insert or update or delete on copydml_test
    for each row execute procedure qqq_trig();
CREATE TRIGGER
create trigger qqqaf after insert or update or delete on copydml_test
    for each row execute procedure qqq_trig();
CREATE TRIGGER
copy (insert into copydml_test (t) values ('f') returning id) to stdout;
psql:sql/copydml.sql:62: NOTICE:  INSERT 8
8
psql:sql/copydml.sql:62: NOTICE:  INSERT 8
copy (update copydml_test set t = 'g' where t = 'f' returning id) to stdout;
psql:sql/copydml.sql:63: NOTICE:  UPDATE 8
8
psql:sql/copydml.sql:63: NOTICE:  UPDATE 8
copy (delete from copydml_test where t = 'g' returning id) to stdout;
psql:sql/copydml.sql:64: NOTICE:  DELETE 8
8
psql:sql/copydml.sql:64: NOTICE:  DELETE 8
drop table copydml_test;
DROP TABLE
drop function qqq_trig();
DROP FUNCTION
